returnn/tf/layers/basic.py

```diff
@@ -6097,16 +6091,61 @@ def __init__(self, condition, true_from, false_from, **kwargs):
     self.false_from = false_from
     if isinstance(condition, bool):
       if condition:
-        self.output = true_from.output.copy("%s_output" % self.name)
+        true_data = true_from.output if isinstance(true_from, LayerBase) else Data.from_tensor(tf.identity(true_from))
+        self.output = true_data.copy("%s_output" % self.name)
+        if self._get_source_allow_inf_in_output(true_from):
+            self.allow_inf_in_output = True
       else:
-        self.output = false_from.output.copy("%s_output" % self.name)
+        false_data = false_from.output if isinstance(false_from, LayerBase) else Data.from_tensor(
+          tf.identity(false_from))
+        self.output = false_data.copy("%s_output" % self.name)
+        if self._get_source_allow_inf_in_output(false_from):
+            self.allow_inf_in_output = True
     else:
       assert isinstance(condition, LayerBase)
       assert condition.output.dtype == "bool"
+      true_data = true_from.output if isinstance(true_from, LayerBase) else Data.from_tensor(tf.identity(true_from))
+      false_data = false_from.output if isinstance(false_from, LayerBase) else Data.from_tensor(tf.identity(false_from))
       self.output.placeholder = where_bc(
         condition=condition.output.copy_compatible_to(self.output, check_dtype=False, check_sparse=False).placeholder,
-        x=true_from.output.copy_compatible_to(self.output).placeholder,
-        y=false_from.output.copy_compatible_to(self.output).placeholder)
+        x=true_data.copy_compatible_to(self.output).placeholder,
+        y=false_data.copy_compatible_to(self.output).placeholder)
+      if self._get_source_allow_inf_in_output(true_from) or self._get_source_allow_inf_in_output(false_from):
+        self.allow_inf_in_output = True
+
+  @classmethod
+  def _transform_source_name(cls, source_name, get_layer):
+    """
+    :param str source_name:
+    :param ((str) -> LayerBase) get_layer: function to get or construct another layer
+    :return: how the input layer should appear in the transformed config dict
+    :rtype: LayerBase|float|int
+    """
+    if isinstance(source_name, str):
+      return get_layer(source_name)
+    return source_name
+
+  @classmethod
+  def _get_source_template(cls, source, name):
```

Comment:
Make the function local to `get_out_data_from_opts`, you are not using it anywhere else.